content/articles/web-apps.md

```diff
@@ -0,0 +1,108 @@
+
+We run all of these components within docker containers to provide reliable portability between installation environments. Typically the front and back ends of the web applications will go into one container, and the research code and interface layer will go into another. We'll run a web proxy in a third container, and there may be additional containers for databases, task queues etc.
+
+
+# Pre-requisites
+
+In order to support this architecture, we require the research code to have the following features so that we can build a stable interface to it:
+
+- Source-controlled in github: we maintain the source code for our applications in github repositories, and we require research code to be available in github too, so that we can develop against it, apply automated tests to it within continuous integration and deploy it using docker.
+
+- Versioned: it's useful to be able surface a version number in the application, of both the research and application code. It's helpful if the research code can provide its own semantic version, but as a minimum we could use a github commit identifier.
+
+- Data sets are available and clearly defined: any data required for the application (data which is used by the research code, or has been generated by it) should be available to automated access for CI and deployment. This could be on a network drive, but it would be preferable if it was in something like a github release. Data should also be versionsed in some way, and its format should be identical between versions, except for known format upgrades.
+
+- Stable: we need to be able to rely on the research code to have a known interface and to run reliably. It's ok for the interface to evolve, as long as we can prepare for that in advance by updating the interface layer, and merging those changes into the research code and interface layer at the same time (and any changes which might be required in the web application too). That way we can have a working main branch in all repositories all the time. It may be appropriate for some interface changes to be managed as evolving metadata so that application code does not need to change e.g. when a parameter is added.
+
+- Tested: It's also good if the research code can have at least a basic suite of tests which run under CI - it's often difficult to test exact output values from research code (e.g. if it is stochastic), but it's useful to at least test that results are being successfully returned in the expected format.
+
+- Has a way of managing long-running processes: if research code invoked by the user from the application is going to take more than a couple of seconds to run, then it will need to run asynchronously of the initial web request. The web application will need some way to check status and progress while the code is running, and to notify the user when it is complete. We can work with the research team to design the best approach for the application.
```

Comment:
```suggestion
- Has a way of managing long-running processes: if research code invoked by the user from the application is going to take more than a couple of seconds to run, then it will need to run asynchronously off the initial web request. The web application will need some way to check status and progress while the code is running, and to notify the user when it is complete. We can work with the research team to design the best approach for the application.
```

Reply:
(Notwithstanding Rich's bigger comment below - Is "from" is better than "of / off" ?

Reply:
Yeah, actually I did mean 'of' rather than 'off', but I think you're right than 'from' might be clearer, though this whole bit is probably going to go anyway!